clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1399,16 +1456,53 @@ class SyclKernelBodyCreator
     return true;
   }
 
-  void enterStruct(const CXXRecordDecl *, FieldDecl *FD) final {
-    MemberExprBases.push_back(BuildMemberExpr(MemberExprBases.back(), FD));
+  bool handleArrayType(FieldDecl *FD, QualType FieldTy) final {
+    if (!cast<ConstantArrayType>(FieldTy)
+             ->getElementType()
+             ->isStructureOrClassType()) {
+      createExprForArray(FD);
+    }
+    return true;
   }
 
-  void leaveStruct(const CXXRecordDecl *, FieldDecl *FD) final {
-    MemberExprBases.pop_back();
+  void enterField(const CXXRecordDecl *RD, FieldDecl *FD) final {
+    if (!FD->getType()->isReferenceType())
```

Comment:
In VisitRecordFields, enterField is called before any checks are done. That's why a reference field might be encountered here. Building a MemberExpr is the only work done "unchecked".
An alternative is to defer the MemberExpr building until after the check, but then the enterField/leaveField calls will need to be inserted in each of the field handlers of the BodyCreator.

Reply:
All members of the kernel lambda end up in this function, but the only action taken is to create an S->M expression in preparation for further processing. S->M is ok to do except for reference types. (A lit test failed for the case where M is of reference type, hence this check). Actual handling of illegal fields will not happen because that occurs after the FieldChecker has told us that the field is valid. For invalid fields the S->M expression is simply popped off the stack by leaveField.